File: networkutils/src/main/java/net/maxsmr/networkutils/monitor/TrafficStatsDelegate.java
package net.maxsmr.networkutils.monitor;

import android.annotation.SuppressLint;
import android.content.Context;
import android.net.TrafficStats;
import org.jetbrains.annotations.NotNull;

import static net.maxsmr.networkutils.monitor.NetworkStatsHelper.isNetworkStatsManagerSupported;


public final class TrafficStatsDelegate {

    public TrafficStatsDelegate() {
        throw new AssertionError("no instances");
    }

    public static boolean shouldUseTrafficStats(int uid) {
        final int callingUid = android.os.Process.myUid();
        return !isNetworkStatsManagerSupported() || callingUid == android.os.Process.SYSTEM_UID || uid > 0 && callingUid == uid;
    }

    @SuppressLint("InlinedApi")
    @SuppressWarnings("ConstantConditions")
    public static long getUidRxBytes(int uid, @NotNull Context context) {
        if (shouldUseTrafficStats(uid)) {
            return TrafficStats.getUidRxBytes(uid);
        } else {
            return new NetworkStatsHelper(context, uid).getPackageRxBytes(context);
        }
    }

    @SuppressLint("InlinedApi")
    @SuppressWarnings("ConstantConditions")
    public static long getUidTxBytes(int uid, @NotNull Context context) {
        if (shouldUseTrafficStats(uid)) {
            return TrafficStats.getUidTxBytes(uid);
        } else {
            return new NetworkStatsHelper(context, uid).getPackageTxBytes(context);
        }
    }


}

File: networkutils/src/main/kotlin/net/maxsmr/networkutils/monitor/TrafficStatsDelegate.kt
package net.maxsmr.networkutils.monitor

import android.annotation.SuppressLint
import android.content.Context
import android.net.ConnectivityManager.TYPE_MOBILE
import android.net.TrafficStats
import android.os.Process
import net.maxsmr.commonutils.getApplicationUid
import net.maxsmr.networkutils.monitor.NetworkStatsWrapper.Companion.isNetworkStatsManagerSupported

fun shouldUseTrafficStats(uid: Int): Boolean {
    val callingUid = Process.myUid()
    return !isNetworkStatsManagerSupported || callingUid == Process.SYSTEM_UID || uid > 0 && callingUid == uid
}

@JvmOverloads
fun getUidRxBytes(packageName: String, context: Context, type: Int = TYPE_MOBILE): Long =
        getUidRxBytes(getApplicationUid(context, packageName) ?: -1, context, type)

@JvmOverloads
@SuppressLint("InlinedApi")
fun getUidRxBytes(uid: Int, context: Context, type: Int = TYPE_MOBILE): Long {
    return if (shouldUseTrafficStats(uid)) {
        TrafficStats.getUidRxBytes(uid)
    } else {
        NetworkStatsWrapper(context, uid).getPackageRxBytes(type)
    }
}

@JvmOverloads
fun getUidTxBytes(packageName: String, context: Context, type: Int = TYPE_MOBILE): Long =
        getUidTxBytes(getApplicationUid(context, packageName) ?: -1, context, type)

@JvmOverloads
@SuppressLint("InlinedApi")
fun getUidTxBytes(uid: Int, context: Context, type: Int = TYPE_MOBILE): Long {
    return if (shouldUseTrafficStats(uid)) {
        TrafficStats.getUidTxBytes(uid)
    } else {
        NetworkStatsWrapper(context, uid).getPackageTxBytes(type)
    }
}
